module/zfs/dmu.c

```diff
@@ -1441,6 +1542,120 @@ dmu_write_uio(objset_t *os, uint64_t object, zfs_uio_t *uio, uint64_t size,
 }
 #endif /* _KERNEL */
 
+static void
+dbuf_cached_bps(spa_t *spa, blkptr_t *bps, uint_t nbps,
+    uint64_t *l1sz, uint64_t *l2sz)
+{
+	int cached_flags;
+
+	if (bps == NULL)
+		return;
+
+	for (size_t blk_off = 0; blk_off < nbps; blk_off++) {
+		blkptr_t *bp = &bps[blk_off];
+		uint64_t *u64p;
+
+		if (BP_IS_HOLE(bp))
+			continue;
+
+		cached_flags = arc_cached(spa, bp);
+		if (cached_flags == 0)
+			continue;
+
+		u64p = (cached_flags & ARC_CACHED_IN_L2) ? l2sz : l1sz;
+		*u64p += BP_GET_LSIZE(bp);
+	}
+}
+
+/*
+ * Estimate DMU object cached size.
+ */
+int
+dmu_object_cached_size(objset_t *os, uint64_t object,
+    uint64_t *l1sz, uint64_t *l2sz)
+{
+	dnode_t *dn;
+	dmu_object_info_t doi;
+	int err = 0;
+
+	*l1sz = *l2sz = 0;
+
+	if (dnode_hold(os, object, FTAG, &dn) != 0)
+		return (0);
+
+	if (dn->dn_maxblkid <= 1) {
+		dnode_rele(dn, FTAG);
+		return (0);
+	}
+
+	dmu_object_info_from_dnode(dn, &doi);
+
+	for (uint64_t off = 0; off < doi.doi_max_offset;
+	    off += dmu_prefetch_max) {
+		/* dbuf_read doesn't prefetch L1 blocks. */
+		dmu_prefetch(os, object, 1, off,
+		    dmu_prefetch_max, ZIO_PRIORITY_NOW);
+	}
+
+	/*
+	 * Hold all valid L1 blocks, asking ARC the status of each BP
+	 * contained in each such L1 block.
+	 */
+	uint_t nbps = bp_span_in_blocks(dn->dn_indblkshift, 1);
+	uint64_t l1blks = dn->dn_maxblkid / nbps;
+
+	rw_enter(&dn->dn_struct_rwlock, RW_READER);
+	for (uint64_t off = 0, max_off = (l1blks + 1) << dn->dn_indblkshift;
+	    off < max_off; off += doi.doi_metadata_block_size) {
+		dmu_buf_impl_t *db = NULL;
+
+		/* Skip L1's that are not present (i.e., sparse) */
+		if (dnode_next_offset(dn, DNODE_FIND_HAVELOCK, &off, 1,
+		    1, 0) != 0) {
```

Comment:
off here seems to be measured at L1 bytes here.  Same time I am not sure about the argument. I suspect it may be measured in L0 bytes.  At least inside it is passed to dbuf_whichblock(), which takes L0 offsets.

Reply:
yep,  `dnode_next_offset()` operates in terms of L0 bytes. I think we should drop using this function to skip since the `dbuf_hold_impl()` will skip it anyways.  Converting the L1 offset to a L0 one just to use `dnode_next_offset()` would make the code more complex than necessary and with little added value. 

In general ,we don't need to iterate over an offset number space and can simplify to iterate over L1 blocks and not need to shift to an offset only to shift it back to a block to call `dbuf_hold_impl()`.

Reply:
Yea, the simplification feels reasonable to me here. I doubt we should often see many-L1-block-sized holes in ZAPs to justify the complexity.  While ZAP shrinking may create some holes, it is not the same as if somebody just create a petabyte sized file with only one block at the end.

Reply:
Thanks.